napari/layers/_shapes_layer/model.py

```diff
@@ -0,0 +1,1620 @@
+            the length of `data` and each element will be applied to each shape
+            otherwise the same value will be used for all shapes.
+        face_color : str | tuple | list
+            If string can be any color name recognized by vispy or hex value if
+            starting with `#`. If array-like must be 1-dimensional array with 3
+            or 4 elements. If a list is supplied it must be the same length as
+            the length of `data` and each element will be applied to each shape
+            otherwise the same value will be used for all shapes.
+        opacity : float | list
+            Opacity of the shapes, must be between 0 and 1.
+        z_index : int | list
+            Specifier of z order priority. Shapes with higher z order are
+            displayed ontop of others. If a list is supplied it must be the
+            same length as the length of `data` and each element will be
+            applied to each shape otherwise the same value will be used for all
+            shapes.
+        """
+        if len(data) == 0:
+            return
+            
+        if np.array(data[0]).ndim == 1:
+            # If a single array for a shape has been passed
+            if shape_type in self.data._types.keys():
+                shape_cls = self.data._types[shape_type]
+                shape = shape_cls(data, edge_width=edge_width,
+                                  edge_color=edge_color, face_color=face_color,
+                                  opacity=opacity, z_index=z_index)
+            else:
+                raise ValueError("""shape_type not recognized, must be one of
+                                 "{'line', 'rectangle', 'ellipse', 'path',
+                                 'polygon'}"
+                                 """)
+            self.data.add(shape)
+        else:
+            # If list of arrays has been passed
+            for i, d in enumerate(data):
+                if type(shape_type) in (np.ndarray, list):
+                    st = shape_type[i]
+                else:
+                    st = shape_type
+                if type(edge_width) in (np.ndarray, list):
+                    ew = edge_width[i]
+                else:
+                    ew = edge_width
+                if type(edge_color) in (np.ndarray, list):
+                    if np.isscalar(edge_color[i]):
+                        ec = edge_color
+                    else:
+                        ec = edge_color[i]
+                else:
+                    ec = edge_color
+                if type(face_color) in (np.ndarray, list):
+                    if np.isscalar(face_color[i]):
+                        fc = face_color
+                    else:
+                        fc = face_color[i]
+                else:
+                    fc = face_color
+                if type(z_index) in (np.ndarray, list):
+                    z = z_index[i]
+                else:
+                    z = z_index
```

Comment:
Ok so what I'd do for all these is:

```python
def ensure_iterable(*args):
    return [ensure_single_iterable(arg) for arg in args]

def ensure_single_iterable(arg):
    if is_iterable(arg):
        return arg
    else:
        return itertools.repeat(arg)
```

(with some fancy logic within `is_iterable`.)

Then:

```python
face_colors, edge_colors, ... = ensure_iterable(face_color, edge_color, ...)
for d, fc, ec, ... in zip(data, face_colors, edge_colors, ...):
```

Reply:
I have tried fixing this, it required some special treatment for colors (which can also be length 3 or 4 arrays. let me know it this looks good now